### PR TITLE
Feat: query params

### PR DESCRIPTION
**Pull request by mcmcgrath13**
* Add query params for
  * date
  * active cluster
  * town
  * stat
  * zoomed
* Keep these in sync with the `ref` variables in the dashboard
* When date or town changes, reset cluster and zoom
* When we're showing old data, add a quick link to see the latest data

### CODE REVIEW

File: src/composables/useQueryParam.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  param: string;
+  ref: Ref;
+  valid: (T) => boolean;
+  push?: boolean;
+  convertInt?: boolean;
+  refField?: string;
+  /* eslint-disable @typescript-eslint/ban-types */
+  paramToVal?: Function;
+  /* eslint-disable @typescript-eslint/ban-types */
+  valToParam?: Function;
```
Comment (eldu) [inline src/composables/useQueryParam.ts:14]: Would something more explicit about the inputs and outputs of the function work for the typescript?

Something like the following?

`(v: string) => string`

File: src/composables/useQueryParam.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+  // On initialization, check if the param already exists.  If it does, validate it, and
+  // if valid, update the ref's value to match the query param.  If there is no param,
+  // initialize the url to match the initial value of the ref.
+  if (route.query[param]) {
+    const paramVal = parseParam(route.query[param]);
+    if (!valid(paramVal)) {
+      throw new Error(`Invalid URL query parameter: ${param} = ${paramVal}`);
+    }
+    setRef(paramVal); // THIS isn't actually updating the ref??
```
Comment (eldu) [inline src/composables/useQueryParam.ts:83]: Is this comment still valid?

File: src/router.ts [hunk 3]
```diff
@@ -39,16 +39,6 @@ const routes: RouteRecordRaw[] = [
 const router = createRouter({
   history: createWebHistory(),
   routes,
-  scrollBehavior(to) {
```
Comment (eldu) [inline src/router.ts:42]: No more scrolling?

**Review by eldu — COMMENTED** [review]
I think I'm still wrapping my head around `useQueryParam`. I'll reread it again in a bit.

**Comment by eldu** [comment]
Also in this moment, should we redirect the user instead to the default page? Or have a redirect link?

Screen Shot 2022-05-11 at 9 29 24 AM

File: src/composables/useQueryParam.ts (hunk 1 above)
Comment (mcmcgrath13) [reply src/composables/useQueryParam.ts:14]: yeah, but we don't really know the types here is the problem.  I tried using generics `(T) => S` but it complained about that.

File: src/composables/useQueryParam.ts (hunk 2 above)
Comment (mcmcgrath13) [reply src/composables/useQueryParam.ts:83]: nope! I shall delete it

**Comment by mcmcgrath13** [comment]
> Also in this moment, should we redirect the user instead to the default page? Or have a redirect link?

The back/home are kind of the redirect link.  We could reset that value to it's otherwise default, but I wasn't sure if it was better to explicitly error on a bad link, or to silently do something else.

File: src/router.ts (hunk 3 above)
Comment (mcmcgrath13) [reply src/router.ts:42]: fixed, and now it scrolls only on path change

**Comment by eldu** [comment]
> > Also in this moment, should we redirect the user instead to the default page? Or have a redirect link?
> 
> The back/home are kind of the redirect link. We could reset that value to it's otherwise default, but I wasn't sure if it was better to explicitly error on a bad link, or to silently do something else.

Oh yes the back and home link! I somehow missed those 🤦

File: src/composables/useQueryParam.ts (hunk 1 above)
Comment (eldu) [reply src/composables/useQueryParam.ts:14]: ahhh i seee, tricky! alright we disable the typescript then

Comment (mcmcgrath13) [reply src/composables/useQueryParam.ts:14]: just eslint, typescript is happy, just poorly informed and able to make a lot of inferences based on this typing

**Comment by maxwellkrieger** [comment]
Tire kicking report 1: Selecting a cluster within a town and then changing the town and selecting a cluster disables Zoom to community button and cluster name does not update

**Review by eldu — COMMENTED** [review]
Hmm i'm trying to reproduce max's error and am also not able to reproduce it? But it happened right?

**Comment by s-bessey** [comment]
I'm also unable to reproduce the error we were seeing yesterday. Weird!

**Comment by mcmcgrath13** [comment]
I wonder if it's a windows thing? @maxwellkrieger you were using chrome, right?

**Review by eldu — APPROVED** [review]
Approve and open an issue for the weird bug to be investigated whenever we can reproduce it?

**Comment by mcmcgrath13** [comment]
created #135 to track the issue with cluster selection

**Pull request merged**